Native matmul

nullplay opened the pull request:
### Implementation of #151705

This PR introduces the initial implementation of native `tl.dot` support in Inductor, with the goal of generating Triton matmul kernels directly—without relying on predefined templates.

To avoid complexity and ease the review process, I plan to split this work into two phases as outlined in #151705:

1. **Basic support** (this PR)
2. **Lazy broadcasting** for optimal performance (future PR)

### Summary of This PR

This PR implements the basic functionality. It does **not** include lazy broadcasting, so the generated kernels may involve explicit `tl.reshape` and `tl.trans` operations before calling `tl.dot`, which introduces some overhead.

### Notable Changes

1. Adds a new config flag: `config.triton.enable_native_matmul`
2. Introduces a new `ops.dot` IR node in Inductor and lowers `aten.mm` and `aten.bmm` to it when native matmul is enabled
3. Enforces tililng suitable for matmul when the native matmul flag is enabled
4. Implements code generation for `ops.dot`
5. Adds Triton autotuning heuristics: for now, I’ve copied the configuration from the existing matmul templates. However, this may not be optimal—it currently takes a long time to tune, and I think there must be a better way to tackle this.


@eellison @jansel @PaulZhang12 @shunting314

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @Lucaskabela @mlazos

nullplay committed:
add ops.dot

nullplay committed:
add ops.dot codegen

nullplay committed:
add heuristics

nullplay committed:
fix

nullplay committed:
lint and fix

jansel commented on torch/_inductor/config.py:
```diff
@@ -1197,6 +1197,9 @@ class triton:
     tile_reductions: bool = False
 
+    # Codegen matmul natively with tl.dot without calling template.
+    enable_native_matmul: bool = False
```
Can you update the PR to turn this on so we can do a full CI run with it enabled to check for bugs?

(After CI is passing we can turn it off again)

nullplay replied:
I’ve just enabled it, but I’m not fully confident about the performance due to the potential overhead from the `reshape` and `transpose` operations. Back in March, Triton compiler didn’t handle these operations efficiently, which resulted in slower performance. To work around this, I had to modify Inductor to emit alternative code—which I had originally planned to include in a follow-up PR.

```python
tmp0 = tl.load(in_ptr0 + (r0_2 + 128 * y0), r0_mask & ymask, eviction_policy='evict_last', other=0.0)
tmp1 = tl.load(in_ptr1 + (x1 + 128 * r0_2), r0_mask & xmask, eviction_policy='evict_last', other=0.0)
tmp2 = tl.dot(tl.reshape(tmp0, [YBLOCK, R0_BLOCK]), tl.trans(tl.reshape(tmp1, [XBLOCK, R0_BLOCK])), allow_tf32=False)
```

jansel commented:
I haven't looked at this super carefully yet, but I kicked off a benchmark run with it enabled here:
https://github.com/pytorch/pytorch/actions/runs/16134785066

It should show up in the dropdown (nullplay_fuse_matmul) here once the jobs finishes:
https://hud.pytorch.org/benchmark/compilers

nullplay committed:
enable native matmul = True

jansel commented:
I approved CI.  The benchmark run is done, looks like there are a number of models that are failing:

TB Inference: https://hud.pytorch.org/benchmark/torchbench/inductor_with_cudagraphs?dashboard=torchinductor&startTime=Tue,%2001%20Jul%202025%2017:58:59%20GMT&stopTime=Tue,%2008%20Jul%202025%2017:58:59%20GMT&granularity=hour&mode=inference&dtype=bfloat16&deviceName=cuda%20(h100)&lBranch=nullplay_fuse_matmul&lCommit=985e1a07e280c19db24a8603aa795c504f0273e7&rBranch=main&rCommit=1586521461c8dc642735466fc143b7d366a858d0

TB Training:
https://hud.pytorch.org/benchmark/torchbench/inductor_with_cudagraphs?dashboard=torchinductor&startTime=Tue,%2001%20Jul%202025%2017:58:59%20GMT&stopTime=Tue,%2008%20Jul%202025%2017:58:59%20GMT&granularity=hour&mode=training&dtype=amp&deviceName=cuda%20(h100)&lBranch=nullplay_fuse_matmul&lCommit=985e1a07e280c19db24a8603aa795c504f0273e7&rBranch=main&rCommit=1586521461c8dc642735466fc143b7d366a858d0


The other benchmark suites can be viewed by selecting "nullplay_fuse_matmul" in the branch dropdown (+ training/inference).

nullplay committed:
fix and disable upcast fp32

jansel commented:
> I fixed a few bugs and pushed the changes again. Could you re-run the CI and performance benchmarks?

Something is odd with CI (in this PR and a few others). I don't see any jobs to approve.

There is also a merge conflict.  Can you rebase?  That will hopefully fix the CI issue.

> I noticed that when doing torch.float16 matmuls, it was automatically upcasting to float32. Disabling config.triton.codegen_upcast_to_fp32 made things faster. I'm not sure what effect this might have on other parts of the code, but I’ve set config.triton.codegen_upcast_to_fp32 = False for now.

This is to match what eager pytorch does for pointwise ops.  Most of those ops are memory bound so the upcast to fp32 doesn't matter for performance.  For matmuls that won't work.  We should modify the upcast logic to not apply to matmuls.   

> Just to confirm—there’s no way for me to trigger the CI myself, right? Or is there a way to run the tests locally on my end?

I just asked to add permissions for you to trigger CI yourself.

You should be able to run tests locally.  Failing tests should print out the repro command and the benchamrks are all in the pytorch/benchmarks folder.

jansel commented:
You should have access to start CI now.  I kicked off another benchmark run here: https://github.com/pytorch/pytorch/actions/runs/16242184585

nullplay committed:
fix and reset upcast_to_fp32

nullplay committed:
fix

nullplay committed:
test accuracy change

nullplay committed:
cleanup pass

nullplay committed:
fix

nullplay committed:
fix

nullplay committed:
fix

nullplay committed:
minor

nullplay committed:
minor

nullplay committed:
restrict dtype

nullplay committed:
disable when MKN = 1

nullplay committed:
fix reduction where

nullplay committed:
fix iteration split on three dimension tiling

nullplay committed:
fix

nullplay committed:
fix

nullplay committed:
lint

nullplay committed:
don't apply reordering when matmul codegen

nullplay committed:
dtype fix

nullplay committed:
minor tests

nullplay committed:
test fix

nullplay committed:
broadcast fix

nullplay committed:
comment

nullplay committed:
fix test

nullplay committed:
fix test

nullplay committed:
fix

nullplay committed:
fix bug on shape prop with dot

nullplay committed:
fix acc shape

nullplay committed:
add load shape prop and fix dot operand issue

nullplay committed:
Blockshapetype get str

nullplay committed:
fix addmm and tests

nullplay committed:
fix bugs

nullplay committed:
fix test

nullplay committed:
fix bugs

nullplay committed:
fix test

nullplay committed:
fix tma

nullplay committed:
fix test

nullplay committed:
fix precision

nullplay committed:
fix

nullplay committed:
cse var

nullplay committed:
test

nullplay committed:
fix

nullplay committed:
fix

nullplay committed:
fix

nullplay committed:
fix

nullplay committed:
for check purpose

nullplay committed:
fix

nullplay committed:
fix bug

nullplay committed:
fix

nullplay committed:
lint

nullplay committed:
lint fix

nullplay committed:
lint

nullplay committed:
lint

nullplay committed:
fix

nullplay commented:
Okay, I think this is now ready for an initial review.

I’ve fixed most of the CI issues with `config.triton.enable_native_matmul=True`. Many of the problems were caused by tests assuming that matmul wouldn’t be generated (e.g., checking for `"extern.mm"` in files). In those cases, I either replaced `extern.mm` with the actual function name or skipped the test when native matmul is enabled. I’m not entirely sure whether I should have fixed tests that aren’t directly related to the core functionality of native matmul, so please let me know if any of those changes should be reverted.

There are still a few CI failures I haven’t been able to resolve — for example, an error in `python pytorch/test/distributed/test_dynamo_distributed.py TestSingleProc.test_graph_split_inductor` that I can’t reproduce locally. However, the majority of the functional bugs related to native matmul should now be resolved.

I also discovered a bug in the Triton compiler while testing TorchBench (`hf_BigBird` and `hf_Reformer`), even though the generated code appears correct. I’ve reported these issues in the Triton repository: [https://github.com/triton-lang/triton/issues/8022](https://github.com/triton-lang/triton/issues/8022) (for `hf_BigBird`) and [https://github.com/triton-lang/triton/issues/8023](https://github.com/triton-lang/triton/issues/8023) (for `hf_Reformer`).

This is my first PR to PyTorch, so the code might be a bit rough or unconventional. Any feedback or reviews would be greatly appreciated. Below is an overview of the implementation details for this feature:

---

### 1. Lower matmul-related aten operations to `ops.dot` in the post-grad FX pass

[https://github.com/pytorch/pytorch/blob/508614818efa4745dca1b11710b116e76d12a93d/torch/\_inductor/fx\_passes/post\_grad.py#L184-L187](https://github.com/pytorch/pytorch/blob/508614818efa4745dca1b11710b116e76d12a93d/torch/_inductor/fx_passes/post_grad.py#L184-L187)

* `aten.addmm` and `aten.baddbmm` are first decomposed into `aten.mm` or `aten.bmm`. Then, `aten.mm` and `aten.bmm` are lowered to `ops.dot`. For example, given `C = aten.mm(A, B)`, this can be rewritten as:

  ```python
  Prod = A.unsqueeze(-1) * B.unsqueeze(0)  # ops.mul -> ops.dot
  C = Prod.sum(dim=1)                      # ops.reduction('sum') -> ops.reduction('dot')
  ```

  Instead of explicitly using `ops.mul` and `ops.reduction("sum")`, these are lowered into a combination of `ops.dot` + `ops.reduction("dot")`. These IR nodes are semantically equivalent to the `ops.mul` + `ops.reduction("sum")` combination but are lowered to `tl.dot` during code generation. See the pass here: [https://github.com/pytorch/pytorch/blob/508614818efa4745dca1b11710b116e76d12a93d/torch/\_inductor/fx\_passes/post\_grad.py#L1624-L1644](https://github.com/pytorch/pytorch/blob/508614818efa4745dca1b11710b116e76d12a93d/torch/_inductor/fx_passes/post_grad.py#L1624-L1644)

* Currently, this only supports matrices where all dimensions are ≥ 2 (no matrix-vector multiplication yet).

---

### 2. Multi-dimensional loop scheduling and tiling

In native matmul (`C[z,y,x] = A[z,y,r] @ B[z,r,x]`), we keep the loop order fixed and do not merge them (loop = `z -> y -> x -> r` for `aten.bmm` and `y -> x -> r` for `aten.mm`). The kernel tile order is fixed accordingly.

* Fixing the loop order of the native matmul node is implemented here:
  [https://github.com/pytorch/pytorch/blob/508614818efa4745dca1b11710b116e76d12a93d/torch/\_inductor/ir.py#L4703-L4717](https://github.com/pytorch/pytorch/blob/508614818efa4745dca1b11710b116e76d12a93d/torch/_inductor/ir.py#L4703-L4717)
* Choosing the kernel tile order `z -> y -> x -> r` is handled here:
  [https://github.com/pytorch/pytorch/blob/508614818efa4745dca1b11710b116e76d12a93d/torch/\_inductor/codegen/simd.py#L2526-L2538](https://github.com/pytorch/pytorch/blob/508614818efa4745dca1b11710b116e76d12a93d/torch/_inductor/codegen/simd.py#L2526-L2538)
* `_split_iteration_ranges` was updated to better support 3D ranges for batched matmul:
  [https://github.com/pytorch/pytorch/blob/508614818efa4745dca1b11710b116e76d12a93d/torch/\_inductor/codegen/simd.py#L710-L747](https://github.com/pytorch/pytorch/blob/508614818efa4745dca1b11710b116e76d12a93d/torch/_inductor/codegen/simd.py#L710-L747)
* Logic in `can_fuse` was updated to prevent fusion between a matmul node and a reduction node when their loop orders differ:
  [https://github.com/pytorch/pytorch/blob/508614818efa4745dca1b11710b116e76d12a93d/torch/\_inductor/codegen/simd.py#L1253-L1272](https://github.com/pytorch/pytorch/blob/508614818efa4745dca1b11710b116e76d12a93d/torch/_inductor/codegen/simd.py#L1253-L1272)

Fixing the loop order is necessary because the code generation for `ops.dot` assumes this order. This restriction isn’t fundamental, so we can investigate reordering in future optimizations. Currently, native matmul reuses the existing fusion heuristics, but better fusion heuristics tailored for native matmul will be an important area of future work.

---

### 3. Code generation for `ops.dot`

The logic for lowering `ops.dot` to `tl.dot` is implemented here:
[https://github.com/pytorch/pytorch/blob/508614818efa4745dca1b11710b116e76d12a93d/torch/\_inductor/codegen/triton.py#L1169-L1390](https://github.com/pytorch/pytorch/blob/508614818efa4745dca1b11710b116e76d12a93d/torch/_inductor/codegen/triton.py#L1169-L1390)

Key steps:

1. **Downcasting for performance:** If data was upcast to fp32 earlier, we downcast back to the original dtype (fp16 or bf16) to maximize throughput. Although surrounding ops may run in fp32, the matmul itself is executed in the original precision for better performance.
2. **Handling non-constant reduction masks:** If the reduction mask is not constant and there were intermediate operations between `tl.load` and `tl.dot`, we zero out values outside the mask using `tl.where(r0_mask, val, 0)` to prevent incorrect results.
3. **Shape alignment for `tl.dot`:** We reshape, transpose, or broadcast tensors to match the required `(Y, R) x (R, X)` shape for `tl.dot`. We avoid using the 3D dot form (`(Z, Y, R) x (Z, R, X)`) due to its poor performance. For batched matmul (`bmm`), we keep `ZBLOCK=1` and call the 2D dot kernel instead.

The shape alignment step currently incurs a 10–50% performance overhead, but this will be addressed in a future PR with lazy broadcasting.

Example:

```python
def f(x, y):
    return x @ (y + 1.0)

x = torch.randn(128,128, device="cuda", dtype=torch.float16)
y = torch.randn(128,128, device="cuda", dtype=torch.float16)
g = torch.compile(f, mode="max-autotune")
```

Generates something like:

```python
tmp0 = tl.load(in_ptr0 + (r0_2 + 128*y0), r0_mask & ymask, eviction_policy='evict_last', other=0.0).to(tl.float32)
tmp1 = tl.load(in_ptr1 + (x1 + 128*r0_2), r0_mask & xmask, eviction_policy='evict_last', other=0.0).to(tl.float32)
tmp2 = 1.0
tmp3 = tmp1 + tmp2

# 1. Downcasting
tmp4 = tmp0.to(tl.float16) 
tmp5 = tmp3.to(tl.float16) 

# 2. Zeroing out outside reduction mask
tmp6 = tl.where(r0_mask, tmp5, 0) 

# 3. Shape alignment for tl.dot
tmp7 = tl.reshape(tmp4, [YBLOCK, R0_BLOCK]) 
tmp8 = tl.trans(tl.reshape(tmp6, [XBLOCK, R0_BLOCK]))

tmp9 = tl.dot(tmp7, tmp8, input_precision="ieee") 
```

---

### 4. Autotuning

The existing `_reduction_config` and `_persistent_reduction_config` in `runtime/triton_heuristic.py` couldn’t generate configurations for native matmul due to the (z, y, x, r) tiling order, which is not supported by the current implementation. so I manually added predefined configs (copied from the default mm template) here:
[https://github.com/pytorch/pytorch/blob/508614818efa4745dca1b11710b116e76d12a93d/torch/\_inductor/runtime/triton\_heuristics.py#L2498-L2591](https://github.com/pytorch/pytorch/blob/508614818efa4745dca1b11710b116e76d12a93d/torch/_inductor/runtime/triton_heuristics.py#L2498-L2591)

I also added a new `native_matmul` key in `triton_meta` to indicate when to use these configs. Right now, \~20 configurations are tested for each kernel, which significantly increases compilation time. We should look into reducing this overhead — for example:

1. Restricting to 5–6 predefined configs by default and using coordinate descent tuning when `max-autotune` is enabled, or
2. Updating `reduction_config` and `persistent_reduction_config` to naturally support native matmul kernels.

nullplay committed:
comment

nullplay commented:
Benchmark(H100, default) : https://hud.pytorch.org/benchmark/compilers?dashboard=torchinductor&startTime=Sun%2C%2031%20Aug%202025%2016%3A38%3A27%20GMT&stopTime=Tue%2C%2002%20Sep%202025%2016%3A38%3A27%20GMT&granularity=hour&mode=inference&dtype=bfloat16&deviceName=cuda%20(h100)&lBranch=nullplay/fuse_matmul&lCommit=4f411816cd398440b526504bff4b4af618335c86&rBranch=main&rCommit=2a5c0785e2f975697fd7bdf1411de6e03dcaa1ef

Failures:
1. `sam` (Torchbench) : accuracy failures
2. `hf_BigBird` (Torchbench) : Triton bug described above
3. `hf_Reformer` (Torchbench) : Triton bug described above
4. `tnt_s_patch16_224` (timm) : Cuda error (grid.z > 65536, too big batch size in bmm)
5. `xcit_large_24_p8_224` (timm) : Cuda error (grid.z > 65536, too big batch size in bmm)

jansel commented on test/inductor/test_max_autotune.py:
Is every test in this file skipped?  Maybe put the skip/patch on the test class so you only need it once rather than for each test.

nullplay replied:
Only some tests that require `FileCheck` are skipped. So let’s keep the skips (we can add new `FileCheck` cases for the newly generated code from `native_matmul` across many tests in this file but that would be a bit time-consuming.)

jansel commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -1508,12 +1706,12 @@ def is_valid_addmm_fusion(match):
         aten.add,
         KeywordArg("inp"),
-        CallFunction(aten.mm, Arg(), Arg()),
+        CallFunction(aten.mm, KeywordArg("mat1"), KeywordArg("mat2")),
```
Will this change the behavior when native_matmul is disabled?

jansel reviewed (changes requested):
Overall looks great!  I made a bunch of comments.

nullplay committed:
atol,rtol

nullplay committed:
fix

nullplay committed:
fixed precision issue

nullplay committed:
add test and add helper in the test

nullplay commented on test/inductor/test_aot_inductor.py:
```diff
@@ -5960,6 +5995,9 @@ def runner_call(*args, **kwargs):
         runner.free_inactive_constant_buffer()
 
+    @unittest.skipIf(
+        config.triton.enable_native_matmul, "K is too large; precision issue"
```
When calling `tl.dot(x, y, input_precision="...")` without specifying `input_precision`, it defaults to `tf32`. I realized this and updated the codegen to explicitly use `tl.dot(x, y, input_precision="ieee")`, which fixed the issue. Now, I removed the skip command in this test.

nullplay commented on test/inductor/test_native_matmul.py:
```diff
@@ -0,0 +1,128 @@
+
+# Make the helper files in test/ importable
+pytorch_test_dir = os.path.dirname(os.path.dirname(os.path.realpath(__file__)))
+sys.path.append(pytorch_test_dir)
```
Removed.

nullplay commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -1471,8 +1492,185 @@ def repl(inp, x1, x2):
+        # Currently only enable native matmul for default indexing
+        # TODO : support block ptr
+        if config.triton.use_block_ptr:
+            return False
```
As commented above, I fixed the code to throw an error when both block_ptr and native_matmul are turned on.

nullplay committed:
Revert unintended third_party changes to match upstream/main

nullplay committed:
enable_native_matmul -> native_matmul

nullplay commented on torch/_inductor/config.py:
```diff
@@ -1290,6 +1290,9 @@ class triton:
     tile_reductions: bool = False
 
+    # Codegen matmul natively with tl.dot without calling template.
+    enable_native_matmul: bool = True
```
Done.

nullplay committed:
fx pass

nullplay commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -1471,8 +1492,185 @@ def repl(inp, x1, x2):
+
 def is_valid_addmm_fusion(match):
-    mat1, mat2 = match.args
+    mat1, mat2 = match.kwargs["mat1"], match.kwargs["mat2"]
```
No, it doesn’t change the behavior; it was a leftover from previous changes. I reverted it back to the original implementation.

nullplay commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -1498,7 +1696,7 @@ def is_valid_addmm_fusion(match):
     CallFunction(
         aten.add,
-        CallFunction(aten.mm, Arg(), Arg()),
+        CallFunction(aten.mm, KeywordArg("mat1"), KeywordArg("mat2")),
```
Reverted it back.

nullplay committed:
cleanup

nullplay commented on torch/_inductor/runtime/triton_heuristics.py:
```diff
@@ -2494,8 +2495,108 @@ def pointwise(
+    ({"z": 1, "x": 128, "y": 128, "r": 64}, 8, 5),
+]
+
+triton_native_persistent_bmm_configs = [
```
Now it is cleaned up.

nullplay commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -745,6 +754,18 @@ def is_valid_mm_plus_mm(match: Match):
     if m1 != m2 or n1 != n2:
         return False
 
+    if config.triton.enable_native_matmul:
```
This is mainly for preventing inductor to call mm_plus_mm template.
When native matmul is turned on, `is_valid_mm_plus_mm` returns False so that it does not call template (i.e., generated natively).


```
# We want to avoid calling this when config.triton.native_matmul is on
@register_lowering_pattern(
    CallFunction(
        aten.add,
        CallFunction(aten.mm, KeywordArg("mat1"), KeywordArg("mat2")),
        CallFunction(aten.mm, KeywordArg("mat3"), KeywordArg("mat4")),
    ),
    extra_check=is_valid_mm_plus_mm,
)
def mm_plus_mm(match: Match, mat1, mat2, mat3, mat4):
    return inductor.kernel.mm_plus_mm.tuned_mm_plus_mm(mat1, mat2, mat3, mat4)
```

nullplay committed:
relax loop ordering

nullplay committed:
simplify the integer indexing

nullplay commented on torch/_inductor/codegen/triton.py:
```diff
@@ -2333,8 +2617,38 @@ def match_block_expr() -> Optional[BlockDescriptorOptions]:
+                #
+                # Instead of broadcasting a dense constant, we use a size-1 scalar constant
+                # to preserve the correct dependency on the [Y, R] axes for tl.dot.
+                expand_str = str([1] * len(self.dense_size_list()))
```
I think it’s always safe to use integer indexing with size-1 tensors (I was just being cautious before). I’ll test the CI by converting all indexing to this style.

nullplay committed:
fix and cleanup dtype

nullplay commented on torch/_inductor/codegen/triton.py:
```diff
@@ -1111,6 +1166,229 @@ def maximum(a, b):
+        orig_a, orig_b = a, b
+
+        # Downcast if we upcasted to fp32 before
+        for node in V.kernel.features.node_schedule:
```
Fixed. We now explicitly insert `ops.to_dtype` before `ops.dot` in the FX pass and remove the type-casting logic from the codegen.

https://github.com/pytorch/pytorch/blob/afcb5a10fc2a18b480a41a6d57445457c2800c11/torch/_inductor/fx_passes/post_grad.py#L1637-L1653

nullplay committed:
tldot dtype

nullplay committed:
fix integer indexing

nullplay committed:
fix

nullplay committed:
fix

nullplay committed:
precision

jansel commented on torch/_inductor/codegen/triton.py:
```diff
@@ -195,6 +195,61 @@ class TritonSymbols:
+        #
+        # get_block_shape(y1) = (YBLOCK,1,1)
+        # get_block_shape(x0-tmp13) = (YBLOCK,XBLOCK,1)
+        from torch._inductor.shape_propagation import get_broadcasted_shape
```
Could we import at top of file?

jansel commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -1471,8 +1492,185 @@ def repl(inp, x1, x2):
     match.replace_by_example(repl, [inp, mat1, mat2])
 
 
+def native_matmul_pass(graph: torch.fx.Graph):
```
I was mostly talking about lowering.py here.  That only operates on a per-op basis though.  Though I think for the most part native matmul only need to touch a single op at a time.  Most of the templates are created directly from lowering.py -- mm_plus_mm is an exception because it spans 3 ops.

For cases where you want to split an op into smaller peices decompositions are the easiest way.  If the decomp returns `NotImplemented` then it is left unchanges.

jansel reviewed (changes requested):
I think this is getting pretty close.  Can you sumarize the remaining test failures?  Fix lints with `lintrunner -a`.

Once this is in a ready to land state we should switch native_matmul off by default.

cc @eellison @shunting314

nullplay committed:
suggested change

nullplay committed:
lint fix

nullplay committed:
fix test

nullplay commented on test/inductor/test_native_matmul.py:
```diff
@@ -0,0 +1,161 @@
+    
+    def test_3mm_add(self):
+        def f(x, y, z, w, r, t):
+            return x @ y + z @ w + r @ t
```
This is the generated code, and yes — it appears that all three accumulators are kept alive until the end.

```python
def triton_red_fused_addmm_mm_0(in_out_ptr0, in_ptr0, in_ptr1, in_ptr2, in_ptr3, in_ptr4, in_ptr5, ynumel, xnumel, r0_numel, YBLOCK : tl.constexpr, XBLOCK : tl.constexpr, R0_BLOCK : tl.constexpr):
    ynumel = 128
    xnumel = 128
    r0_numel = 128
    rnumel = r0_numel
    RBLOCK: tl.constexpr = R0_BLOCK
    yoffset = tl.program_id(1) * YBLOCK
    yindex = yoffset + tl.arange(0, YBLOCK)[:, None, None]
    ymask = yindex < ynumel
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[None, :, None]
    xmask = xindex < xnumel
    r0_base = tl.arange(0, R0_BLOCK)[None, None, :]
    rbase = r0_base
    y0 = yindex
    x1 = xindex
    _tmp6 = tl.full([YBLOCK, XBLOCK], 0, tl.float32)
    for r0_offset in range(0, r0_numel, R0_BLOCK):
        r0_index = r0_offset + r0_base
        r0_mask = r0_index < r0_numel
        roffset = r0_offset
        rindex = r0_index
        r0_2 = r0_index
        tmp0 = tl.load(in_ptr0 + (r0_2 + 128*y0), r0_mask & ymask, eviction_policy='evict_last', other=0.0)
        tmp1 = tl.load(in_ptr1 + (x1 + 128*r0_2), r0_mask & xmask, eviction_policy='evict_last', other=0.0)
        tmp2 = tl.reshape(tmp0, [YBLOCK, R0_BLOCK])
        tmp3 = tl.trans(tl.reshape(tmp1, [XBLOCK, R0_BLOCK]))
        tmp4 = tl.dot(tmp2, tmp3, input_precision="ieee")
        tmp5 = tl.broadcast_to(tmp4, [YBLOCK, XBLOCK])
        tmp7 = _tmp6 + tmp5
        _tmp6 = tmp7
    tmp6 = _tmp6[:,:,None]
    _tmp14 = tl.full([YBLOCK, XBLOCK], 0, tl.float32)
    _tmp22 = tl.full([YBLOCK, XBLOCK], 0, tl.float32)
    for r0_offset in range(0, r0_numel, R0_BLOCK):
        r0_index = r0_offset + r0_base
        r0_mask = r0_index < r0_numel
        roffset = r0_offset
        rindex = r0_index
        r0_2 = r0_index
        tmp8 = tl.load(in_ptr2 + (r0_2 + 128*y0), r0_mask & ymask, eviction_policy='evict_last', other=0.0)
        tmp9 = tl.load(in_ptr3 + (x1 + 128*r0_2), r0_mask & xmask, eviction_policy='evict_last', other=0.0)
        tmp16 = tl.load(in_ptr4 + (r0_2 + 128*y0), r0_mask & ymask, eviction_policy='evict_last', other=0.0)
        tmp17 = tl.load(in_ptr5 + (x1 + 128*r0_2), r0_mask & xmask, eviction_policy='evict_last', other=0.0)
        tmp10 = tl.reshape(tmp8, [YBLOCK, R0_BLOCK])
        tmp11 = tl.trans(tl.reshape(tmp9, [XBLOCK, R0_BLOCK]))
        tmp12 = tl.dot(tmp10, tmp11, input_precision="ieee")
        tmp13 = tl.broadcast_to(tmp12, [YBLOCK, XBLOCK])
        tmp15 = _tmp14 + tmp13
        _tmp14 = tmp15
        tmp18 = tl.reshape(tmp16, [YBLOCK, R0_BLOCK])
        tmp19 = tl.trans(tl.reshape(tmp17, [XBLOCK, R0_BLOCK]))
        tmp20 = tl.dot(tmp18, tmp19, input_precision="ieee")
        tmp21 = tl.broadcast_to(tmp20, [YBLOCK, XBLOCK])
        tmp23 = _tmp22 + tmp21
        _tmp22 = tmp23
    tmp14 = _tmp14[:,:,None]
    tmp22 = _tmp22[:,:,None]
    tmp24 = 0.0
    tmp25 = tmp6 + tmp24
    tmp26 = tmp25 + tmp14
    tmp27 = tmp26 + tmp24
    tmp28 = tmp27 + tmp22
    tl.debug_barrier()
    tl.store(in_out_ptr0 + (x1 + 128*y0), tmp28, xmask & ymask)
```

shunting314 replied:
Thanks for showing the generated code. This is slightly different to what I image. Any idea why 2 loops is generated rather than a single loop that does 3 tl.dot in the body?

nullplay commented on torch/_inductor/codegen/simd.py:
```diff
@@ -1190,6 +1254,28 @@ def can_fuse(self, node1, node2):
                     rnumel2,
                 )
+
+            if reduction_can_fuse and node1.is_native_matmul():
```
I found that swapping `node1` and `node2` and calling this function again at the very end ensures that it also handles the case where `node2` is the native matmul. (Correct me if I’m wrong.)

```python
        # swap args to hit the case above
        return self.can_fuse_horizontal(node2, node1)
```

nullplay committed:
fix

nullplay committed:
relax the tolerance

nullplay committed:
fix blockshape in tma addmm

nullplay committed:
revert torchbench precision

nullplay committed:
remove fx pass and do lowering

nullplay committed:
better can_fuse for native matmul

nullplay committed:
fix lowering logic

nullplay committed:
remove native matmul from postgrad

nullplay committed:
don't do native matmul on unbacked symbols

nullplay committed:
addmm baddbmm zero annihilation

eellison reviewed (commented):
Very cool ! 

I would recommend we land this with it off by default, and document what the follow ups are

nullplay committed:
turn native matmul off

nullplay committed:
fix lint

nullplay committed:
remove unnecessary changes

nullplay committed:
fix indexing in store and store_reduction and triton_meta.get
